### PR TITLE
fix: use decodeURIComponentSafe for files whose name containing `%`

### PR DESCRIPTION
**Pull request by ooker777**
see: [Why does decodeURIComponent('%') lock up my browser?](https://stackoverflow.com/a/54310080/3416774)

### Check List

- [x] Have you read the
      [CODE OF CONDUCT](https://github.com/lumeland/lume/blob/master/CODE_OF_CONDUCT.md)
- [x] Have you read the document
      [CONTRIBUTING](https://github.com/lumeland/lume/blob/master/CONTRIBUTING.md)
  - [x] One pull request per feature. If you want to do more than one thing,
        send multiple pull request.
  - [ ] Write tests.
  - [x] Run deno `fmt` to fix the code format before commit.
  - [x] Document any change in the `CHANGELOG.md`.

### CODE REVIEW

**Comment by oscarotero** [comment]
Thanks!
Can you give an example of a filename that needs this?

**Comment by oscarotero** [comment]
Okay, after the conversation on Discord, I can see now the bug.
But it should be fixed in the `server.ts` file, because the problem is on serving the file, not generating it.

Can you modify the code [here](https://github.com/lumeland/lume/blob/main/core/server.ts#L131) to this?

```js
const pathname = decodeURIComponent(url.pathname.replace(/%(?![0-9][0-9a-fA-F]+)/g, '%25'));
```

**Comment by oscarotero** [comment]
and you can document the change in the changelog file.
I'm not testing the server class, so no problem if you don't test it.

**Comment by ooker777** [comment]
hmm, this seems to work fine, but using the wiki theme the menu text doesn't show up. Any idea?

**Review by oscarotero — CHANGES_REQUESTED** [review]

File: core/server.ts [hunk 1]
```diff
@@ -128,7 +128,9 @@ export async function serveFile(
   request: Request,
 ): Promise<Response> {
   const url = new URL(request.url);
-  const percentEscapedPathname = url.pathname.replace(/%(?![0-9][0-9a-fA-F]+)/g, '%25');
+  const percentEscapedPathname = url.pathname.replace(/%(?![0-9a-fA-F]+)/g, '%25');
```
[the remark this answers is not in the document]
Comment (ooker777) [reply core/server.ts:131]: Yes. This is to include unicode characters. The letter `Ó`, for example, is encoded as `%C3%93` in UTF-8

File: core/fs.ts [hunk 2]
```diff
@@ -118,8 +118,15 @@ export default class FS {
...
+    let dirEntries
 
-    for (const dirEntry of Deno.readDirSync(dirPath)) {
+    try {
+      dirEntries = Deno.readDirSync(dirPath);
+    } catch {
+      console.error(`Error: The system cannot find the path at ${dirPath}.`)
+      Deno.exit()
+    }
+    for (const dirEntry of dirEntries) {
```
Comment (oscarotero) [inline core/fs.ts:129]: Out of curiosity, what's the purpose of this? In what scenario the `Deno.readDirSync` can fail?
Just want to understand the motivation of this.

Comment (ooker777) [reply core/fs.ts:129]: Oh, it's unrelated to this PR. It should be on another PR, but my git skills was not fluent enough to split it to another branch yet still existing in the code for me to test, so I put it there and then forgot to take it out when pushed. I'll revert that when I work on my computer later.

It's just to give a nicer error when the user specify a source folder but it doesn't exist.

**Pull request merged**